File: services/server/src/safe_socket/safe_socket.py
import socket

def recv_all(socket: socket.socket, size):
    bytesRead = 0
    total = bytes()
    while bytesRead < size:
        message = socket.recv(size - bytesRead)
        total += message
        bytesRead += len(message)
        if len(message) == 0:
            return total
    return total


def send_all(socket: socket.socket, bytes):
    bytesWritten = 0
    while bytesWritten < len(bytes):
        n = socket.send(bytes[bytesWritten:])
        bytesWritten += n
    return bytesWritten

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"io"
)

func SendAll(socket io.Writer, bytes []byte) error {
	bytesWritten := 0
	for bytesWritten < len(bytes) {
		n, err := socket.Write(bytes[bytesWritten:])
		if err != nil {
			return err
		}
		bytesWritten += n
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	total := make([]byte, 0, size)
	bytesRead := 0
	for bytesRead < size {
		buff := make([]byte, size-bytesRead)
		n, err := socket.Read(buff)
		total = append(total, buff[:n]...)
		bytesRead += n
		if err != nil {
			if err == io.EOF {
				return total, nil
			}
			return nil, err
		}
	}
	return total, nil
}
